app/assets/javascripts/discourse/app/components/translation.gjs

```diff
@@ -0,0 +1,193 @@
+ * This component allows for complex i18n scenarios where you need to embed interactive
+ * components within translated text.
+ *
+ * If you don't require this functionality, use the standard i18n() function.
+ *
+ * @component Translation
+ *
+ * @template Usage example:
+ * ```gjs
+ * // Translation key: "some.translation.key" = "Welcome, %{username}! The date is %{shortdate}!"
+ * <Translation
+ *   @scope="some.translation.key"
+ *   @options={{hash shortdate=shortDate}}
+ * >
+ *   <:placeholders as |Placeholder|>
+ *     <Placeholder @name="username">
+ *       <UserLink @user={{user}}>{{user.username}}</UserLink>
+ *     </Placeholder>
+ *   </:placeholders>
+ * </Translation>
+ * ```
+ *
+ * @param {String} scope - The i18n translation key to use
+ * @param {Object} [options] - Hash of options to pass to the i18n function for string interpolation
+ */
+export default class Translation extends Component {
+  /**
+   * A map of placeholder keys to their unique identifiers.
+   *
+   * @type {Map<String, String>}
+   */
+  _placeholderKeys = new Map();
+
+  /**
+   * A map of placeholder keys to their corresponding DOM elements.
+   *
+   * @type {Map<String, HTMLElement>}
+   */
+  _placeholderElements = new Map();
+
+  /**
+   * A map of placeholder keys to their appearance in the translation string.
+   *
+   * @type {Map<String, String>}
+   */
+  _placeholderAppearance = new Map();
+
+  /**
+   * Tracks which placeholders have been rendered.
+   *
+   * @type {Array<String>}
+   */
+  _renderedPlaceholders = [];
+
+  /**
+   * Processes the translation string and returns an array of text segments and
+   * placeholder elements that can be rendered in the template.
+   *
+   * @returns {Array<String|HTMLElement>} Array of text segments and placeholder elements
+   */
+  get textAndPlaceholders() {
+    const optionsArg = this.args.options || {};
+
+    // Find all of the placeholders in the string we're looking at.
+    const message = I18n.findTranslationWithFallback(this.args.scope, {
+      ...optionsArg,
+    });
+    this._placeholderAppearance = I18n.findPlaceholders(message);
+
+    // We only need to keep the placeholders that aren't being handled by those passed in @options.
+    Object.keys(optionsArg).forEach((stringPlaceholder) =>
+      this._placeholderAppearance.delete(stringPlaceholder)
+    );
+
+    this._placeholderAppearance.forEach((_, placeholderName) => {
+      this._placeholderKeys.set(
+        placeholderName,
+        `__PLACEHOLDER__${placeholderName}__${uniqueId()}__`
+      );
+      this._placeholderElements.set(
+        placeholderName,
+        document.createElement("span")
+      );
+    });
+
+    const text = i18n(this.args.scope, {
+      ...Object.fromEntries(this._placeholderKeys),
+      ...optionsArg,
+    });
+
+    // Bail early if there were no placeholders we need to handle.
+    if (this._placeholderAppearance.size === 0) {
+      return [text];
+    }
```

Comment:
I think we can move this up to before the `this._placeholderAppearance.forEach` loop.

Reply:
I've left it where it is, since the `this._placeholderAppearance.forEach` loop shouldn't run when this condition is triggered, but moving it up would mean having multiple places where `i18n()` is called.